Added docker support

HarlamovBuldog opened the pull request.

HarlamovBuldog committed:
Added docker support

pisarukv commented on Dockerfile.multistage:
```diff
@@ -0,0 +1,24 @@
+# Install the package
+RUN go install -v ./...
+# Build the Go app
+RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o /go/bin/app .
```
do you understand all the flags?

pisarukv commented on Dockerfile.multistage:
```diff
@@ -0,0 +1,24 @@
+RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o /go/bin/app .
+
+######## Start a new stage from scratch #######
+FROM alpine:latest  
```
you can use scratch. And don't use latest, always bind to some specific version. Only in that way you will have reproducible builds

pisarukv commented on Dockerfile.multistage:
```diff
@@ -0,0 +1,24 @@
+# Start from golang v1.10.4 base image
```
That comments looks like how to tutorial comments. But in fact they are absolutely useless and doesn't give any information. Write comments only in difficult, not obvious places. It will keep you code clean.

pisarukv commented on Dockerfile.multistage:
```diff
@@ -0,0 +1,24 @@
+# Copy all from your project inside workdir of docker image
+COPY . .
+# Download dependencies
+RUN go get -d -v ./...
```
Lets add module or dep with vendor folder. Keep vendor folder committed.

pisarukv commented on Dockerfile.multistage:
```diff
@@ -0,0 +1,24 @@
+# Copy the Pre-built binary file from the previous stage
+COPY --from=builder /go/bin/app .
+# Starting bundled binary file
+CMD [ "sh", "-c", "./app https://dev.1c-bitrix.ru/examples/3_type.zip sfafasfafas https://wordpress.org/wordpress-4.4.2.zip https://wordpress.org/latest.zip" ]
```
why you have links inside dockerfile?

HarlamovBuldog replied:
just for testing purposes and quick launch. I mean in order to run code on fly right after running container

pisarukv replied:
it's not a best idea to keep test data in docker file

HarlamovBuldog replied:
got it. Do i need to just write `CMD ["/bin/sh"]` then?

pisarukv replied:
CMD [ "sh", "-c", "./app"]

HarlamovBuldog committed:
Comments were reworked. Unnecessary commands were deleted. Some commands were simplified.

HarlamovBuldog committed:
Create config.yml

HarlamovBuldog committed:
Merge pull request #3 from HarlamovBuldog/add-circleci-config

Create config.yml

HarlamovBuldog committed:
Build command was rewritten in dockerfile

It was made to fix problem when you can't run main app in resulting container cause not all dependencies has been built

HarlamovBuldog committed:
Add certificates

Without certificates in resulting container program will not be able do download anything from web. To fix this problem they were added.

HarlamovBuldog committed:
Minor changes in dockerfile

Root container was changed to golang:alpine because it's more lightweight and it suits well our needs.
Executing file name was changed to "main"

HarlamovBuldog committed:
CMD was changed to ENTRYPOINT in dockerfile

We need to pass values when running "docker run image_name [args]" . ENTRYPOINT provides this functionality.

pisarukv commented on Dockerfile.multistage:
```diff
@@ -0,0 +1,27 @@
+# Start a new build stage
+FROM scratch
+
+# Copy sertificates from previous stage in order to make download from web work
```
```suggestion
# Copy certificates from previous stage in order to make download from web work
```

HarlamovBuldog committed:
Correct spelling error in Dockerfile.multistage

Co-Authored-By: Vadim Pisaruk <pisarukvadim@gmail.com>
